### PR TITLE
Binary HttpBody

### PR DESCRIPTION
**Pull request by ghik**
So far it was only possible to send textual bodies encoded as UTF-8.

### CODE REVIEW

File: rest/.jvm/src/test/scala/io/udash/rest/examples/UserApi.scala [hunk 1]
```diff
@@ -2,13 +2,53 @@ package io.udash
 package rest.examples
```
Comment (ghik) [inline rest/.jvm/src/test/scala/io/udash/rest/examples/UserApi.scala]: Please ignore these changes. These are accidentally commited changes, but harmless.

**Comment by mkej** [comment]
LGTM

File: rest/.jvm/src/main/scala/io/udash/rest/RestServlet.scala [hunk 2]
```diff
@@ -70,9 +99,12 @@ class RestServlet(handleRequest: RawRest.HandleRequest, handleTimeout: FiniteDur
     restResponse.headers.entries.foreach {
       case (name, PlainValue(value)) => response.addHeader(name, value)
     }
-    restResponse.body.forNonEmpty { (content, mimeType) =>
-      response.setContentType(s"$mimeType;charset=utf-8")
-      response.getWriter.write(content)
+    restResponse.body.nonEmptyOpt.foreach { neBody =>
```
Comment (ddworak) [inline rest/.jvm/src/main/scala/io/udash/rest/RestServlet.scala]: Unrelated comment: HTTP responses should always set content length in servlets afair (at least for Jetty), regardless of how underlying implementation might handle it.

File: rest/src/main/scala/io/udash/rest/raw/JsonValue.scala [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+package rest.raw
+
+import com.avsystem.commons.serialization.GenCodec
+import com.avsystem.commons.serialization.json.RawJson
+
+/**
+  * Value used as encoding of [[io.udash.rest.Body Body]] parameters of
+  * [[io.udash.rest.JsonBody JsonBody]] methods. Wrapped value MUST be a valid JSON.
+  */
+case class JsonValue(value: String) extends AnyVal
```
Comment (ddworak) [inline rest/src/main/scala/io/udash/rest/raw/JsonValue.scala]: final

File: rest/src/main/scala/io/udash/rest/raw/PlainValue.scala [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import io.udash.utils.URLEncoder
+
+/**
+  * Value used as encoding of [[io.udash.rest.Path Path]], [[io.udash.rest.Header Header]] and
+  * [[io.udash.rest.Query Query]] parameters as well as [[io.udash.rest.Body Body]] parameters of
+  * [[io.udash.rest.FormBody FormBody]] methods.
+  *
+  * Wrapped string MUST NOT be URL-encoded.
+  */
+case class PlainValue(value: String) extends AnyVal
```
Comment (ddworak) [inline rest/src/main/scala/io/udash/rest/raw/PlainValue.scala]: final

File: rest/src/main/scala/io/udash/rest/raw/PlainValue.scala [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  final val FormKVPairSep = "&"
+
+  def encodeQuery(query: Mapping[PlainValue]): String =
+    query.entries.iterator.map { case (name, PlainValue(value)) =>
+      s"${URLEncoder.encode(name, spaceAsPlus = true)}$FormKVSep${URLEncoder.encode(value, spaceAsPlus = true)}"
+    }.mkString(FormKVPairSep)
+
+  def decodeQuery(queryString: String): Mapping[PlainValue] = {
+    val builder = Mapping.newBuilder[PlainValue]
+    queryString.split(FormKVPairSep).iterator.filter(_.nonEmpty).map(_.split(FormKVSep, 2)).foreach {
```
Comment (ddworak) [inline rest/src/main/scala/io/udash/rest/raw/PlainValue.scala]: fold with the builder as zero?

File: rest/src/main/scala/io/udash/rest/raw/RestRequest.scala [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  */
+final class HttpMethod(implicit enumCtx: EnumCtx) extends AbstractValueEnum
+object HttpMethod extends AbstractValueEnumCompanion[HttpMethod] {
+  final val GET, HEAD, POST, PUT, DELETE, CONNECT, OPTIONS, TRACE, PATCH: Value = new HttpMethod
+}
+
+/**
+  * Aggregates serialized path, header and query parameters of an HTTP request.
+  */
+case class RestParameters(
```
Comment (ddworak) [inline rest/src/main/scala/io/udash/rest/raw/RestRequest.scala]: final

File: rest/src/main/scala/io/udash/rest/raw/RestRequest.scala [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      path ::: method.applyPathParams(otherParameters.path),
+      headers ++ otherParameters.headers,
+      query ++ otherParameters.query
+    )
+}
+object RestParameters {
+  final val Empty = RestParameters()
+}
+
+case class HttpErrorException(code: Int, payload: OptArg[String] = OptArg.Empty, cause: Throwable = null)
```
Comment (ddworak) [inline rest/src/main/scala/io/udash/rest/raw/RestRequest.scala]: final

File: rest/src/main/scala/io/udash/rest/raw/RestRequest.scala [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  extends RuntimeException(s"HTTP ERROR $code${payload.fold("")(p => s": $p")}", cause) with NoStackTrace {
+  def toResponse: RestResponse = RestResponse.plain(code, payload)
+}
+
+case class RestRequest(method: HttpMethod, parameters: RestParameters, body: HttpBody)
```
Comment (ddworak) [inline rest/src/main/scala/io/udash/rest/raw/RestRequest.scala]: final

File: rest/src/main/scala/io/udash/rest/raw/RestResponse.scala [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+package rest.raw
+
+import com.avsystem.commons._
+import com.avsystem.commons.misc.ImplicitNotFound
+import com.avsystem.commons.rpc.{AsRaw, AsReal}
+
+import scala.annotation.implicitNotFound
+import scala.util.Failure
+
+case class RestResponse(code: Int, headers: IMapping[PlainValue], body: HttpBody) {
```
Comment (ddworak) [inline rest/src/main/scala/io/udash/rest/raw/RestResponse.scala]: final

File: rest/src/main/scala/io/udash/rest/raw/RestResponse.scala [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    HttpErrorException(code, body.textualContentOpt.toOptArg)
+  def ensureNonError: RestResponse =
+    if (code >= 200 && code < 300) this else throw toHttpError
+}
+
+object RestResponse extends RestResponseLowPrio {
+  def plain(status: Int, message: OptArg[String] = OptArg.Empty): RestResponse =
+    RestResponse(status, IMapping.empty, HttpBody.plain(message))
+
+  class LazyOps(private val resp: () => RestResponse) extends AnyVal {
```
Comment (ddworak) [inline rest/src/main/scala/io/udash/rest/raw/RestResponse.scala]: final

File: rest/src/main/scala/io/udash/rest/raw/RestResponse.scala [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    def recoverHttpError: RestResponse = try resp() catch {
+      case e: HttpErrorException => e.toResponse
+    }
+  }
+  implicit def lazyOps(resp: => RestResponse): LazyOps = new LazyOps(() => resp)
+
+  implicit class AsyncOps(private val asyncResp: RawRest.Async[RestResponse]) extends AnyVal {
```
Comment (ddworak) [inline rest/src/main/scala/io/udash/rest/raw/RestResponse.scala]: Stay consistent with `LazyOps` in using or not using implicit classes.

File: rest/src/main/scala/io/udash/rest/raw/JsonValue.scala (hunk 3 above)
Comment (ghik) [reply rest/src/main/scala/io/udash/rest/raw/JsonValue.scala]: Value classes are always final.

File: rest/src/main/scala/io/udash/rest/raw/PlainValue.scala (hunk 4 above)
Comment (ghik) [reply rest/src/main/scala/io/udash/rest/raw/PlainValue.scala]: Value classes are always final.

File: rest/src/main/scala/io/udash/rest/raw/PlainValue.scala (hunk 5 above)
Comment (ghik) [reply rest/src/main/scala/io/udash/rest/raw/PlainValue.scala]: It's a mutable builder, I prefer to use imperative style with it.

File: rest/src/main/scala/io/udash/rest/raw/RestRequest.scala (hunk 7 above)
Comment (ghik) [reply rest/src/main/scala/io/udash/rest/raw/RestRequest.scala]: I could actually see some benefit in this being non-final.

File: rest/src/main/scala/io/udash/rest/raw/RestResponse.scala (hunk 10 above)
Comment (ghik) [reply rest/src/main/scala/io/udash/rest/raw/RestResponse.scala]: stahp

File: rest/src/main/scala/io/udash/rest/raw/RestResponse.scala (hunk 11 above)
Comment (ghik) [reply rest/src/main/scala/io/udash/rest/raw/RestResponse.scala]: No, `LazyOps` is not `implicit` only because value classes can't wrap by-name params.

**Pull request merged**